### ISSUE TITLE
Multiple Issues With Variant 2.1 (x86) Unpacker

### ISSUE THREAD

**Issue by atom0s** [atom0s/Steamless#102]
The variant 2.1 (x86) unpacker has several bugs. Because of how this variant works, it makes use of the SteamDRMP.dll file to handle various parts of the unpacking/decryption. The current way that Steamless is handling this causes the offsets being pulled to not be correct all the time. While some v2.1 files work fine, others fail entirely because the data being used is wrong.

Additional patterns would be ideal to confirm which offsets are being used for which parts of the unpacking.

The main issue(s) stem from:

  - https://github.com/atom0s/Steamless/blob/master/Steamless.Unpacker.Variant21.x86/Main.cs#L319
  - https://github.com/atom0s/Steamless/blob/master/Steamless.Unpacker.Variant21.x86/Main.cs#L415
  - https://github.com/atom0s/Steamless/blob/master/Steamless.Unpacker.Variant21.x86/Main.cs#L559

There needs to be a better manner of handling these offsets that are pulled from the DLL to ensure all files making use of this variant work correctly.

The following issues are affected by this problem: #16, #46, #101

**Comment by JollyWizard** [comment]
Issues with Obulis, which reports v2.1 as well.

After unpacking with default options, it reports that steam must be running.  Tried with various options and get same result or it launches and the process terminates immediately. I ran it from the CLI and no errors were reported before the process dies.

**Comment by atom0s** [comment]
> Issues with Obulis, which reports v2.1 as well.
> 
> After unpacking with default options, it reports that steam must be running. Tried with various options and get same result or it launches and the process terminates immediately. I ran it from the CLI and no errors were reported before the process dies.

If Steamless worked on the file, then that kind of error is due to the game also using the Steamworks API. (steam_api.dll / steam_api64.dll) Steamless does not remove that.

**Comment by JollyWizard** [comment]
> > Issues with Obulis, which reports v2.1 as well.
> > After unpacking with default options, it reports that steam must be running. Tried with various options and get same result or it launches and the process terminates immediately. I ran it from the CLI and no errors were reported before the process dies.
> 
> If Steamless worked on the file, then that kind of error is due to the game also using the Steamworks API. (steam_api.dll / steam_api64.dll) Steamless does not remove that.

Thanks for clarifying. I wasn't familiar with the symptoms.

**Comment by atom0s** [comment]
Parts of these issues have been fixed with the latest version of Steamless. (3.1.0.5)

**Comment by rec32** [comment]
I got an exception with game "Iron Front: Liberation 1944" (Iron Front: Digital War Edition) which uses SteamStub 2.1

> Index out of range. The index must be a positive number and its size must not exceed the size of the collection

No options helped against this, I'm using 3.1.0.5

**Comment by atom0s** [comment]
> I got an exception with game "Iron Front: Liberation 1944" (Iron Front: Digital War Edition) which uses SteamStub 2.1
> 
> > Index out of range. The index must be a positive number and its size must not exceed the size of the collection
> 
> No options helped against this, I'm using 3.1.0.5

Hello, can you upload the exe for that game somewhere and link it here?

**Comment by rec32** [comment]
[Full game](https://drive.google.com/uc?export=download&id=1vu2o8BwpWBbjZLfnYbCEYwkimpsn5b9F) / [exe](https://pixeldrain.com/u/wgnbFBZj)

**Comment by MapleHinata** [comment]
I'm currently encountering this issue with the latest depot of Final Fantasy XIII (AppId 292120), with the default options, the latest version of Steamless reports an failure when attempting to decrypt the .text section, with the "Use Experimental Features" option enabled, Steamless outputs a broken executable.

Here is the executable.
[ffxiiiimg.zip](https://github.com/user-attachments/files/16687395/ffxiiiimg.zip)

**Comment by mirh** [comment]
Also add #113

**Comment by R1C3R21** [comment]
I can confirm the same issue does happen with RPG Maker MV from version 1.0.0 to 1.3.5.

**Comment by Ruchian** [comment]
> I can confirm the same issue does happen with RPG Maker MV from version 1.0.0 to 1.3.5.

This is incorrect. I have RPG Maker MV myself. It is not packed with any SteamStub. The reason why it still launches Steam and won't run without is due to the steam_api.dll link. I am not sure what Steam is used for as it doesn't provide any features to this program.

**Comment by R1C3R21** [comment]
> > I can confirm the same issue does happen with RPG Maker MV from version 1.0.0 to 1.3.5.
> 
> This is incorrect. I have RPG Maker MV myself. It is not packed with any SteamStub. The reason why it still launches Steam and won't run without is due to the steam_api.dll link. I am not sure what Steam is used for as it doesn't provide any features to this program.

Well, in those versions there is still no Steam integration, so it is effectively used for DRM. What I don't understand is that "link" part you do mention here.

**Comment by Ruchian** [comment]
> > > I can confirm the same issue does happen with RPG Maker MV from version 1.0.0 to 1.3.5.
> > 
> > 
> > This is incorrect. I have RPG Maker MV myself. It is not packed with any SteamStub. The reason why it still launches Steam and won't run without is due to the steam_api.dll link. I am not sure what Steam is used for as it doesn't provide any features to this program.
> 
> Well, in those versions there is still no Steam integration, so it is effectively used for DRM. What I don't understand is that "link" part you do mention here.

I was wondering if I should explain this or not, but I decided to. "Link" is extremely simplified.
What I mean is that the program itself is build with instructions that look for functions provided by the steam_api.dll in a way that makes the program stop running once it notices it cannot use the steam_api.dll's methods to contact the Steam Network, or be notified that the Steam Client is running offline mode. This is a form of DRM, as you mentioned, as it will not run without the other program running and connecting to the Steam servers. 

What Steamless attempts to remove is not this function. Instead, it is something else Steam provides to Steam Partners.
https://partner.steamgames.com/doc/features/drm
Steam provides partners the option to "wrap" their executable up inside a package (I'm trying to be simple here). You could say that this package surrounds the executable, and must be opened up before the true executable can run. It is kinda like a 'launcher'.

Steamless attempts to remove this, internalized steam client launch-wrapper, so that the real executable can just run without needing to go through this wrapper's instructions first.

What you're looking for basically is something that either removes the instructions to contact the steam client through steam_api.dll from the RPGMV.exe file, or something that simply reports to the game or this dll that the client is running or something.

**Comment by aishadreamer** [comment]
Already made a post in issue [#113](https://github.com/atom0s/Steamless/issues/113), but I'll mention it again here for clarity's sake. The original Dead Island Riptide fails to unpack with the following exception:
`[Steamless] Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index`
I also uploaded the game's executable in that thread.

**Comment by ghost** [comment]
Happens with the Steam version of NBA 2K14
Detects SteamStub Variant 2.1 and has an exception when attempting to unpack

**Comment by go2tom42** [comment]
Happens with the steam version of MLB 2k12

Detects SteamStub Variant 2.1 With exception "Index was out of range. Must be non-negative and less the size of Parameter name: Index"

EXE attached 

[mlb2k12.zip](https://github.com/user-attachments/files/26485123/mlb2k12.zip)

**Comment by AITYunivers** [comment]
> Index out of range. The index must be a positive number and its size must not exceed the size of the collection

I fixed this by just forcing the CodeSectionIndex to 0 if it's not encrypted. It's ugly, and probably breaks very specific games, but it works for me and that's all that matters.

<img width="1317" height="256" alt="Image" src="https://github.com/user-attachments/assets/a6c320c3-1801-46f2-8bf1-80327a139140" />
[Steamless.Unpacker.Variant21.x86] Main.cs Line 400
